command/agent/exec/exec.go

```diff
@@ -0,0 +1,288 @@
+		<-ctx.Done()
+		return nil
+	}
+
+	managerConfig := ctmanager.ManagerConfig{
+		AgentConfig: s.config.AgentConfig,
+		Namespace:   s.config.Namespace,
+		LogLevel:    s.config.LogLevel,
+		LogWriter:   s.config.LogWriter,
+	}
+
+	runnerConfig, runnerConfigErr := ctmanager.NewConfig(managerConfig, s.config.AgentConfig.EnvTemplates)
+	if runnerConfigErr != nil {
+		return fmt.Errorf("template server failed to runner generate config: %w", runnerConfigErr)
+	}
+
+	// we leave in "dry" mode, as there's no files
+	// we will get the env var rendered contents from incoming events
+	var err error
+	s.runner, err = manager.NewRunner(runnerConfig, true)
+	if err != nil {
+		return fmt.Errorf("template server failed to create: %w", err)
+	}
+
+	idMap := s.runner.TemplateConfigMapping()
+	lookupMap := make(map[string][]*ctconfig.TemplateConfig, len(idMap))
+	for id, ctmpls := range idMap {
+		for _, ctmpl := range ctmpls {
+			tl := lookupMap[id]
+			tl = append(tl, ctmpl)
+			lookupMap[id] = tl
+		}
+	}
+	s.lookupMap = lookupMap
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.runner.Stop()
+			s.procLock.Lock()
+			if s.proc != nil {
+				s.proc.Stop()
+			}
+			s.procStarted = false
+			s.proc.Unlock()
+			return nil
+		case token := <-incomingVaultToken:
+			if token != *latestToken {
+				s.logger.Info("exec server received new token")
+
+				s.runner.Stop()
+				*latestToken = token
+				ctv := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token:           latestToken,
+						ClientUserAgent: pointerutil.StringPtr(useragent.AgentTemplatingString()),
+					},
+				}
+
+				runnerConfig = runnerConfig.Merge(&ctv)
+				var runnerErr error
+				s.runner, runnerErr = manager.NewRunner(runnerConfig, true)
+				if runnerErr != nil {
+					s.logger.Error("template server failed with new Vault token", "error", runnerErr)
+					continue
+				}
+				go s.runner.Start()
+			}
+
+		case err := <-s.runner.ErrCh:
+			s.logger.Error("template server error", "error", err.Error())
+			s.runner.StopImmediately()
+
+			// Return after stopping the runner if exit on retry failure was specified
+			if s.config.AgentConfig.TemplateConfig != nil && s.config.AgentConfig.TemplateConfig.ExitOnRetryFailure {
+				return fmt.Errorf("template server: %w", err)
+			}
+
+			s.runner, err = manager.NewRunner(runnerConfig, true)
+			if err != nil {
+				return fmt.Errorf("template server failed to create: %w", err)
+			}
+			go s.runner.Start()
+		case <-s.runner.TemplateRenderedCh():
+			// A template has been rendered, figure out what to do
+			s.logger.Debug("template rendered")
+			events := s.runner.RenderEvents()
+
+			// events are keyed by template ID, and can be matched up to the id's from
+			// the lookupMap
+			if len(events) < len(s.lookupMap) {
+				// Not all templates have been rendered yet
+				continue
+			}
+
+			// assume the renders are finished, until we find otherwise
+			doneRendering := true
+			envVarToContents := map[string]string{}
```

Comment:
I think this can also be a slice of strings or { key, value } structs. This way the order will be preserved

Reply:
also I believe the underlying system expects this values as comma sep `[]string`. If this is used to dedupe before turning into a slice I totally understand but maybe we can move to helper.